geopandas/tools/sjoin.py

```diff
@@ -186,6 +180,351 @@ def find_contains(a1, a2):
+    search_radius : int or float, default None
+        Restricts search to a certain radius.
+        This can significantly speed up execution.
+        If using a tight search radius, it is recommended that you do not use the
+          max_search_neighbors option as it will not improve performance.
+        If you are not using search_radius, consider using max_search_neighbors
+          to speed up execution.
+    nearest_distances: bool, default False
+        If True, report the distance for each match in new column
+          named "nearest_distances".
+    """
+    # ------------------------------ CHECK INPUTS ------------------------------
+    allowed_hows = ["left", "right", "inner"]
+    if how not in allowed_hows:
+        raise ValueError(
+            '`how` was "%s" but is expected to be in %s' % (how, allowed_hows)
+        )
+
+    _basic_checks(left_df, right_df)
+
+    # ----------------- RENAME INDEXES FOR RTREE COMPATIBILITY -----------------
+    (
+        left_df,
+        right_df,
+        left_index_name,
+        right_index_name,
+        index_left,
+        index_right,
+    ) = _rename_indexes(left_df, right_df, lsuffix, rsuffix)
+
+    # ------------------------ COMPUTE SPATIAL JOIN ----------------------------
+
+    # get spatial index
+    tree_idx = right_df.sindex
+
+    # validate max_search_neighbors and search_radius params
+    if max_search_neighbors is not None:
+        assert max_search_neighbors >= 1 and isinstance(
+            max_search_neighbors, int
+        ), "max_search_neighbors must be an integer and >= 1"
+        # warn about using rtree < 0.9.4 and max_search_neighbors option
+        # see https://github.com/Toblerity/rtree/pull/141
+
+        if [int(i) for i in RTREE_VERSION.split(".")] < [0, 9, 4]:
+            warn(
+                "Using an rtree version < 0.9.4 may cause inconsistent "
+                "results when using max_search_neighbors. Consider using a "
+                "large number or max_search_neighbors=None."
+            )
+    if search_radius is not None:
+        assert search_radius >= 0, "search_radius must be >=0"
+
+    def query_index(geo_in_l):
+        """
+        Queries the spatial index to filter results.
+        """
+        # restric by radius
+        if search_radius is not None:
+            if geo_in_l.geom_type == "Point" and search_radius == 0:
+                # applying .buffer(0) to a point returns an empty geometry
+                geo_search = geo_in_l.coords[0]
+            else:  # polygon or buffer > 0
+                geo_search = geo_in_l.buffer(search_radius).bounds
+            in_radius = list(tree_idx.intersection(geo_search))
+        else:
+            in_radius = right_df.index
+        # find neighbors
+        if max_search_neighbors is not None:
+            neighbors = list(
+                tree_idx.nearest(geo_in_l.bounds, num_results=max_search_neighbors)
```

Comment:
I still not fully understand how this works. How does it measure distance? I suppose it is based on bounding boxes only, so it might yield incorrect results if num_results is small, or not?

Reply:
Exactly right. Just like the existing `sjoin`, the `rtree` spatial index operates on bounding boxes. The spatial index results are then "double-checked" by checking the distance using the `shapely` `distance` method (in the existing `sjoin`, `intersects` is used instead).

Overlapping bounding boxes are not a problem (all results at the same distance are checked). The only edge case I found is the following:
![74060768-83516a00-499f-11ea-9a94-06a203484c3c](https://user-images.githubusercontent.com/1755071/75118755-a3d12380-5631-11ea-8a81-04a3fb3d0ec4.png)

Here, if `max_search_neighbors=1`, the red geometry would match but not the blue. If `max_search_neighbors>1`, both would be checked and there would be no issue. I think that for _most_ datasets even `max_search_neighbors=1` would not cause any issues, and even then, it can be mitigated by using `max_search_neighbors=2`. I think despite this issue, the option is still very valuable: to my knowledge it is the only way to perform a 'nearest' spatial join without restricting the search to a certain radius.

I tried to make this clear to the users in the docs:
```
note:: using a very small number (<5) can cause unexpected results
            due to the implementation of the underlaying spatial index.
          Geometries are converted to bounding boxes, thus results for the
            nearest geometry may unexpected. Using a larger number mitigates
            this, but using too large of a number will provide little performance
            improvement.
```

Reply:
Cool. So it is a distance to the edge of the box. Okay. I'll try to properly check the rest of the code in following days. Thanks.